### PR TITLE
A little gift for you ❤️ - Add Dependencies Injection 

### PR DESCRIPTION
**Pull request by leohubert**
Thank you  :) 
I was too lazy to migrate .js into .ts <3

So as a thank you I share you my DI (Dependencies Injection) :) 
With that your code more cleanly

If you want to know more about DI, i remain available.

### CODE REVIEW

**Comment by matheusgrieger** [comment]
This might be have an obvious answer, but after reading your code (very nice concept, btw) I couldn't see for myself, why did you change the `target` in `tsconfig.json` from `esnext` to `es5`?

**Comment by leohubert** [comment]
I do not know why exactly, I had problems with esnext and Inversify http://inversify.io

File: tsconfig.json [hunk 1]
```diff
@@ -1,11 +1,18 @@
 {
   "compilerOptions": {
-    "target": "esnext",
+    "target": "es5",
```
Comment (fwouts) [inline tsconfig.json]: I wonder if you could put it back to `esnext` if you set `lib` to `["esnext", "dom"]`?

File: tsconfig.json [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
     "moduleResolution": "node",
-    "strict": true,
-    "noImplicitReturns": true,
-    "noUnusedLocals": true,
-    "experimentalDecorators": true
+    "lib": [
+      "es6",
+      "dom"
+    ],
+    "types": [
```
Comment (fwouts) [inline tsconfig.json:11]: What's your reasoning for specifying `types` explicitly? I believe this means that @types/* packages are no longer automatically included, which would be a surprise for people who are starting with this boilerplate (see https://www.typescriptlang.org/docs/handbook/tsconfig-json.html).

File: tsconfig.json [hunk 3]
```diff
@@ -1,11 +1,18 @@
 {
   "compilerOptions": {
-    "target": "esnext",
+    "target": "es5",
+    "strict": true,
```
Comment (fwouts) [inline tsconfig.json:4]: `"strict": true` is a good change, but conceptually it's independent from setting up injection. Would you be OK splitting it into a separate PR?

File: package.json [hunk 4]
```diff
@@ -106,6 +108,7 @@
     "vue-template-compiler": "^2.4.2",
     "webpack": "^3.5.2",
     "webpack-dev-server": "^2.7.1",
+    "webpack-env": "^0.8.0",
```
Comment (fwouts) [inline package.json]: What is this needed for?

File: src/main/entities.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import LoggerService from "./services/logger.service";
+import AppService from "./services/app.service";
+
+
+/**
+ * List of classes what you want to be autoInjectable
```
Comment (fwouts) [inline src/main/entities.ts]: Sorry for being OCD, would you mind adding a period at the end of each sentence?

File: src/main/entities.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * Like angular DI
+ *
+ */
+export default [
+    LoggerService,
```
Comment (fwouts) [inline src/main/entities.ts]: For consistency with the rest of the codebase, please keep indentation to two spaces (here and in other places).

File: src/main/entities.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import LoggerService from "./services/logger.service";
+import AppService from "./services/app.service";
+
+
+/**
+ * List of classes what you want to be autoInjectable
+ *
+ * Like angular DI
```
Comment (fwouts) [inline src/main/entities.ts]: Replace this with a link to http://inversify.io

File: src/main/index.ts [hunk 8]
```diff
@@ -1,58 +1,42 @@
...
-    height: 563,
-    useContentSize: true,
-    width: 1000
-  });
-
-  mainWindow.loadURL(winURL);
+/**
+ * Create the DI Container
+ *
+ * @type {Container}
```
Comment (fwouts) [inline src/main/index.ts]: I don't think JSDoc is necessary here. A plain comment is more lightweight and the type comment is redundant with the code:

```
// Create a container for dependency injection.
let container = new Container();
```

File: src/main/index.ts [hunk 9]
```diff
@@ -1,58 +1,42 @@
...
+ */
+let container = new Container();
 
-  mainWindow.on("closed", () => {
-    mainWindow = null;
-  });
+/**
+ * Bind all entities into the DI Container
+ */
+for(let entity of entities) {
```
Comment (fwouts) [inline src/main/index.ts]: Space after "for"

(we need to set up TSLint!)

File: src/main/index.ts [hunk 10]
```diff
@@ -1,58 +1,42 @@
...
+ *
+ * @type {Container}
+ */
+let container = new Container();
 
-  mainWindow.on("closed", () => {
-    mainWindow = null;
-  });
+/**
+ * Bind all entities into the DI Container
```
Comment (fwouts) [inline src/main/index.ts]: Replace "DI" with "dependency injection", since not everyone is familiar with the concept and that's much easier to run a Google search for.

File: src/main/services/app.service.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
+import {injectable} from 'inversify';
+import {app, BrowserWindow} from 'electron';
+
+/**
```
Comment (fwouts) [inline src/main/services/app.service.ts]: JSDoc isn't the right type of comment here. Instead, please add:
```
// TODO: Consider a cleaner alternative.
```

File: src/main/services/app.service.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+const winUrl = process.env.NODE_ENV === 'development'
+    ? `http://localhost:9080`
+    : `file://${__dirname}/index.html`;
+
+@injectable()
+export default class AppService {
+
+    private mainWindow: BrowserWindow | null = null;
+
+    constructor() {
```
Comment (fwouts) [inline src/main/services/app.service.ts]: Is this empty constructor necessary?

File: src/main/services/logger.service.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import {injectable} from "inversify";
+
+/**
+ * Logger class for beautiful logs <3
```
Comment (fwouts) [inline src/main/services/logger.service.ts]: This file probably belongs in a separate PR as well. Right now, it feels quite incomplete (maybe it should be using `chalk` under the hood so different types of messages appear in different colours?).

**Review by fwouts — CHANGES_REQUESTED** [review]
Thanks, and sorry for the delay. I've added a bunch of comments, but this looks good overall!

File: tsconfig.json (hunk 2 above)
Comment (leohubert) [reply tsconfig.json:11]: For that, i just followed http://inversify.io/

File: package.json (hunk 4 above)
Comment (leohubert) [reply package.json]: Oh sorry, it was for testing, not useful here.

File: tsconfig.json (hunk 3 above)
Comment (leohubert) [reply tsconfig.json:4]: Can you tell me what is PR x)

File: tsconfig.json (hunk 1 above)
Comment (leohubert) [reply tsconfig.json]: I will try

File: src/main/services/logger.service.ts (hunk 13 above)
Comment (leohubert) [reply src/main/services/logger.service.ts]: It was the idea ^^

File: tsconfig.json (hunk 3 above)
Comment (matheusgrieger) [reply tsconfig.json:4]: @MrDarkSkil PR = Pull Request, exactly what you've made here.

File: src/main/services/app.service.ts (hunk 12 above)
Comment (leohubert) [reply src/main/services/app.service.ts]: I forget to remove it x) Sorry

File: tsconfig.json (hunk 3 above)
Comment (leohubert) [reply tsconfig.json:4]: Ah ok ok :)

File: tsconfig.json (hunk 1 above)
Comment (leohubert) [reply tsconfig.json]: It's good now for you ?

File: src/main/entities.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import LoggerService from "./services/logger.service";
+import AppService from "./services/app.service";
+
+/**
+ * List of classes what you want to be autoInjectable.
```
Comment (fwouts) [inline src/main/entities.ts:5]: Nit: "that" you want to be autoInjectable.

File: src/main/services/logger.service.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import {injectable} from "inversify";
+
+/**
+ * Logger class for beautiful logs <3.
+ */
+@injectable()
+export default class LoggerService {
```
Comment (fwouts) [inline src/main/services/logger.service.ts:7]: Would you mind adjusting the indentation here too?

File: tsconfig.json (hunk 1 above)
Comment (fwouts) [reply tsconfig.json]: Thanks! I see that this still says "es5". What error do you run into if you change "target" to "esnext"?

File: tsconfig.json [hunk 16]
```diff
@@ -1,11 +1,18 @@
...
     "moduleResolution": "node",
-    "strict": true,
-    "noImplicitReturns": true,
-    "noUnusedLocals": true,
-    "experimentalDecorators": true
+    "lib": [
+      "esnext",
+      "dom"
+    ],
+    "types": [
```
Comment (fwouts) [inline tsconfig.json:11]: What happens if you remove this "types" section in package.json?

**Comment by fwouts** [comment]
This boilerplate isn't necessary anymore (see README). Thanks for your contribution!
